pysat/utils/registry.py

```diff
@@ -15,117 +15,305 @@
-    with open(os.path.join(pysat.pysat_dir, 'user_modules.txt'), 'w') as f:
-        for mod in pysat.user_modules:
-            f.write(mod + '\n')
+    """Store registered pysat.Instrument user modules to disk"""
 
+    with open(os.path.join(pysat.pysat_dir, 'user_modules.txt'), 'w') as fopen:
+        for platform in pysat.user_modules:
+            for name in pysat.user_modules[platform]:
+                # instrument module string
+                inst_mod = pysat.user_modules[platform][name]
+                # format for storage
+                # platform, name, instrument module
+                out = ' '.join((platform, name, inst_mod, '\n'))
+                # store
+                fopen.write(out)
 
-def register(module_name):
-    """Registers a user module by name, returning the loaded module
+
+def register(module_names):
+    """Registers a user pysat.Instrument module by name
+
+    Enables instantiation of a third-party Instrument
+    module using
+    ::
+        inst = pysat.Instrument(platform, name, tag=tag, sat_id=sat_id)
 
     Parameters
     -----------
-    module_name : string
-        specify package name and instrument module
-        examples:
-            my.package.name.my_instrument
-            my.pckage.name.myInstrument
+    module_names : list-like of str
+        specify package name and instrument modules
 
-    Returns
-    --------
-    Updates the user module registry specified in
-    pysat_dir/user_module.txt
+    Raises
+    ------
+    ValueError
+        If platform and name associated with module_name(s) are currently
+        registered
 
+    Warnings
+    --------
+    Registering a module that contains code other than pysat instrument
+    files could result in unexpected consequences.
 
-    Notes
-    ------
+    Note
+    ----
     Modules should be importable using
+    ::
         from my.package.name import my_instrument
 
     Module names do not have to follow the pysat platform_name naming
     convection.
 
-    Warning: Registering a module that contains code other than
-    pysat instrument files could result in unexpected consequences.
+    Current registered modules bay be found at
+    ::
+        pysat.user_modules
+
+    which is stored as a dict of dicts keyed by platform and name.
 
     Examples
-    ---------
-    from pysat import Instrument, user_modules
-    from pysat.utils import registry
+    --------
+    ::
+        from pysat import Instrument, user_modules
+        from pysat.utils import registry
+
+        registry.register('my.package.name.myInstrument')
+
+        testInst = Instrument(platform, name)
 
-    registry.register('my.package.name.myInstrument')
-    assert 'my.package.name.myInstrument' in user_modules
+    """
+
+    for module_name in module_names:
+        # first, ensure module string directs to something importable
+        try:
+            inst_module = importlib.import_module(module_name)
+        except Exception:
```

Comment:
Do we know what type of Exception is expected?

Reply:
My intent with this try: except is verify we can import the supplied input and provide some context to any error if one occurs. The raise a couple lines below preserves the original error and stack trace for the user to deal with.  

No errors are allowed on an import to ensure, to the extent possible, that the registry only has good packages written to it.

Reply:
There use to be more of these in earlier versions. It started with my logging the error in a variety of places, since, logging is good. But most of the entries would end up like,
```
pysat ERROR: 'platname1'
```
which wasn't all that helpful. I tried a few versions but non seemed to really fit the bill. This one was a residual. After thinking about my response above, it really doesn't do very much.

Reply:
I also have a similar modification over in _instrument.py, so worth sorting out our style.
```
                # not a native pysat.Instrument
                try:
                    mod = user_modules[self.platform][self.name]
                    inst = importlib.import_module(mod)
                except Exception:
                    estr = ' '.join(('pysat was either unable to locate the',
                                     'module in the registry or',
                                     'successfully import the module',
                                     'for platform:', self.platform,
                                     'and name:', self.name))
                    logger.error(estr)
                    raise

```

Reply:
The notion of friendlier error message came up on a recent review that threw a KeyError when I forgot a tag.

Reply:
Well, the reason I don't like this is that there are too many things it will catch.  I like catching exceptions and giving friendlier messages, but specifying which one would be more useful.  For example, you could have:
```
try:
    mod = user_modules[self.platform][self.name]
    inst = importlib.import_module(mod)
except KeyError as kerr:
    estr = "unknown platform or name supplied to user modules: {:} or {:} not in {:}".format(self.platform, self.name, user_modules.__repr__())
    logger.error(estr)
    raise KeyError(kerr)
except ImportError as ierr:
    estr = "unable to locate or import module for platform {:}, name {:}".format(self.platform, self.name)
    logger.error(estr)
    raise ImportError(ierr)
```

Then, if something else breaks we'll still get it in the intended form and can fix or catch it as appropriate.